### PR TITLE
nx-adsp plugin that generates angular app

### PR DESCRIPTION
**Pull request by jonathanweyermann**
nx-adsp plugin that generates angular app

instructions are also in the nx-adsp README.md

# to test once deployed

1. Create your Nx workspace using `npx create-nx-workspace my-workspace`
2. Install plugin: `npm i -D @abgov/nx-adsp`
3. Generate a quick start `npx g @abgov/nx-adsp:angular-app my-app`

# to test before it's merged

Stuff inside [] needs to be replaced with your own names

1. Generate a temp folder with a local nx-adsp using `nx run nx-adsp-e2e:e2e`
2. Generate a new project somewhere with `npx create-nx-workspace`, following the prompts
3. add the following to the dev dependencies in package.json inside the new workspace (this assumes you an angular plugin)

    "@abgov/nx-adsp": "[location-of-nx-tools]/dist/packages/nx-adsp",
    "@nrwl/angular": "^12.0.8",

4. run npm install
5. generate the plugin using `npx nx g @abgov/nx-adsp:[pluginName] my-ang-app --tenant tenantName`
6. run npm install
7. serve the plugin using `nx serve [pluginName]
-------

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -27,11 +27,13 @@
   },
   "private": true,
   "dependencies": {
-    "@nrwl/devkit": "~11.5.2"
+    "@angular-devkit/schematics-cli": "^0.1102.10",
+    "@nrwl/devkit": "12.0.8"
```
Comment (tingzuge-dio) [inline package.json:31]: Let's keep the @nrwl dependencies at 11.5.2 if possible; I think generally they have to be kept at the same version.

File: e2e/ang-e2e/tests/ang.spec.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
+import {
```
Comment (tingzuge-dio) [inline e2e/ang-e2e/tests/ang.spec.ts:1]: workspace.json doesn't look like it has an ang-e2e project? 

Please merge the specs into e2e/nx-adsp-e2e, and remove the extraneous files.

File: packages/nx-adsp/package.json [hunk 3]
```diff
@@ -13,7 +13,8 @@
     "@abgov/nx-dotnet": "^1.0.0",
     "@abgov/nx-oc": "^1.3.0",
     "@nrwl/express": "^11.5.2",
-    "@nrwl/react": "^11.5.2"
+    "@nrwl/react": "^11.5.2",
+    "@nrwl/angular": "^12.0.8"
```
Comment (tingzuge-dio) [inline packages/nx-adsp/package.json:17]: keep this consistent with the other @nrwl lib version.

File: packages/nx-adsp/src/generators/angular-app/angular-app.ts [hunk 4]
```diff
@@ -0,0 +1,162 @@
+import {
+  addDependenciesToPackageJson,
+  addProjectConfiguration,
+  formatFiles,
+  generateFiles,
+  getWorkspaceLayout,
+  installPackagesTask,
```
Comment (tingzuge-dio) [inline packages/nx-adsp/src/generators/angular-app/angular-app.ts:7]: return a callback that runs this task to install the dependencies added with addDependenciesToPackageJson

File: packages/nx-adsp/README.md [hunk 5]
```diff
@@ -7,4 +7,28 @@ The plugin includes generators for application, service, and fullstack solution
...
+## To add a plugin to @abgov/nx-adsp
+
+nx g @nrwl/nx-plugin:plugin [pluginName] --importPath .
+move the code generated into packages/ns-adsp and make changes in the plugin directory you've just generated
+
+## To test the plugin locally
+
+1. Generate a temp folder with a local nx-adsp using `nx run nx-adsp-e2e:e2e`
+2. Generate a new project somewhere with `npx create-nx-workspace`, following the prompts
+3. add the following to the dev dependencies in package.json inside the new workspace (this assumes you an angular plugin)
```
Comment (tingzuge-dio) [inline packages/nx-adsp/README.md:26]: I think you can actually use: npm i -D [location-of-nx-tools]/dist/packages/nx-adsp to update package.json.

File: packages/nx-adsp/src/generators/angular-app/files/src/polyfills.ts__tmpl__ [hunk 6]
```diff
@@ -0,0 +1,66 @@
+/**
```
Comment (tingzuge-dio) [inline packages/nx-adsp/src/generators/angular-app/files/src/polyfills.ts__tmpl__:1]: curious about this file... is the polyfills not included in the base nrwl angular generator, and/or are we modifying it in some way?

Comment (jonathanweyermann) [reply packages/nx-adsp/src/generators/angular-app/files/src/polyfills.ts__tmpl__:1]: yeah, if we don't add 


 import '@angular/localize/init';


to the bottom of the file, we get an error - I think the angular package in ui-components has the same file line at the bottom

File: packages/nx-adsp/src/generators/angular-app/angular-app.ts (hunk 4 above)
Comment (jonathanweyermann) [reply packages/nx-adsp/src/generators/angular-app/angular-app.ts:7]: This caused some sort of dependency error - that said, I think it worked fine if I manually ran npm install after this was run, but it was kind of confusing. I can add it in.

Comment (jonathanweyermann) [reply packages/nx-adsp/src/generators/angular-app/angular-app.ts:7]: yeah, If I do this,

I get the following when I serve the app

> nx run my-ang-app:serve
Cannot find module 'webpack'
Require stack:
- C:\Users\Jonathan\src\stuff26\node_modules\@angular-devkit\build-webpack\src\webpack\index.js
- C:\Users\Jonathan\src\stuff26\node_modules\@angular-devkit\build-webpack\src\index.js
- C:\Users\Jonathan\src\stuff26\node_modules\@angular-devkit\build-angular\src\dev-server\index.js
- C:\Users\Jonathan\src\stuff26\node_modules\@angular-devkit\architect\node\node-modules-architect-host.js
- C:\Users\Jonathan\src\stuff26\node_modules\@angular-devkit\architect\node\index.js
- C:\Users\Jonathan\src\stuff26\node_modules\@nrwl\tao\src\commands\ngcli-adapter.js
- C:\Users\Jonathan\src\stuff26\node_modules\@nrwl\tao\src\commands\run.js
- C:\Users\Jonathan\src\stuff26\node_modules\@nrwl\tao\index.js
- C:\Users\Jonathan\src\stuff26\node_modules\@nrwl\cli\lib\run-cli.js

and If I install webpack I get

Jonathan@DESKTOP-VQKLAV0 MINGW64 ~/src/stuff26 (master)
$ npm install webpack

added 37 packages, removed 1 package, and audited 1875 packages in 10s

133 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities

Jonathan@DESKTOP-VQKLAV0 MINGW64 ~/src/stuff26 (master)
$ nx serve my-ang-app

> nx run my-ang-app:serve
The target entry-point "@abgov/angular-components" has missing dependencies:
 - @angular/cdk/overlay


———————————————————————————————————————————————

>  NX   ERROR  Running target "my-ang-app:serve" failed

  Failed tasks:

  - my-ang-app:serve

and when I install 
 npm install @angular/cdk

$ nx serve my-ang-app --verbose

> nx run my-ang-app:serve --verbose
Cannot read property 'add' of undefined

File: packages/nx-adsp/src/generators/angular-app/angular-app.ts [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+      {
+        ...options,
+        project: normalizedOptions.projectName
+      }
+    );
+  }
+
+  // return () => {
+  //   installPackagesTask(host);
+  // }
```
Comment (jonathanweyermann) [inline packages/nx-adsp/src/generators/angular-app/angular-app.ts:165]: The diff if I did add this

$ diff package.json ../stuff29/package.json
2c2
<   "name": "stuff27",
---
>   "name": "stuff29",
26c26,27
<     "help": "nx help"
---
>     "help": "nx help",
>     "postinstall": "ngcc --properties es2015 browser module main"
30,36c31,41
<     "core-js": "^3.6.5",
<     "document-register-element": "1.13.1",
<     "react": "17.0.2",
<     "react-dom": "17.0.2",
<     "regenerator-runtime": "0.13.7",
<     "styled-jsx": "3.4.4",
<     "tslib": "^2.0.0"
---
>     "@angular/animations": "^11.2.0",
>     "@angular/common": "^11.2.0",
>     "@angular/compiler": "^11.2.0",
>     "@angular/core": "^11.2.0",
>     "@angular/forms": "^11.2.0",
>     "@angular/platform-browser": "^11.2.0",
>     "@angular/platform-browser-dynamic": "^11.2.0",
>     "@angular/router": "^11.2.0",
>     "rxjs": "~6.6.3",
>     "tslib": "^2.0.0",
>     "zone.js": "^0.10.2"
39,42c44,53
<     "@babel/core": "7.12.13",
<     "@babel/preset-env": "7.12.13",
<     "@babel/preset-react": "7.12.13",
<     "@babel/preset-typescript": "7.12.13",
---
>     "@abgov/angular-components": "1.0.1",
>     "@abgov/core-css": "^0.7.56",
>     "@abgov/nx-adsp": "file:../stuff2/nx-tools/dist/packages/nx-adsp",
>     "@angular-devkit/build-angular": "~0.1102.0",
>     "@angular-eslint/eslint-plugin": "~4.2.0",
>     "@angular-eslint/eslint-plugin-template": "~4.2.0",
>     "@angular-eslint/template-parser": "~4.2.0",
>     "@angular/compiler-cli": "^11.2.0",
>     "@angular/language-service": "^11.2.0",
>     "@nrwl/angular": "^12.1.0",
48d58
<     "@nrwl/react": "12.1.0",
50d59
<     "@nrwl/web": "12.1.0",
52d60
<     "@testing-library/react": "11.2.6",
55,57d62
<     "@types/react": "17.0.3",
<     "@types/react-dom": "17.0.3",
<     "@types/styled-jsx": "2.2.8",
60d64
<     "babel-jest": "26.2.2",
66,69c70
<     "eslint-plugin-import": "2.22.1",
<     "eslint-plugin-jsx-a11y": "6.4.1",
<     "eslint-plugin-react": "7.23.1",
<     "eslint-plugin-react-hooks": "4.2.0",
---
>     "html-webpack-plugin": "~4.5.2",
70a72,73
>     "jest-preset-angular": "8.4.0",
>     "oidc-client": "~1.11.5",

File: packages/nx-adsp/src/generators/angular-app/angular-app.ts (hunk 4 above)
Comment (tingzuge-dio) [reply packages/nx-adsp/src/generators/angular-app/angular-app.ts:6]: seems to be due to --legacy-peer-deps; good to omit the task for now.

**Pull request merged**